[flang] Fix execute_command_line cmdstat is not set when error occurs

yiwu0b11 opened the pull request:
Fixes: https://github.com/llvm/llvm-project/issues/92929

yiwu0b11 committed:
[flang] Fix execute_command_line cmdstat is not set when error occurs

Fixes: https://github.com/llvm/llvm-project/issues/92929

DanielCChen commented:
Nit:
With the patch, the code in PR #92929 now issues 
```
> a.out
 No error!
cat: no: No such file or directory
 exitstat= 1
 cmdstat= 3
 Invalid command line
 ```
 
 Both `exitstat` and `cmdstat` are correct now, but the error message `Invalid command line` seems not quite accurate. 
 For instance, if I changed the command to a random `fakecommand`, it will give the same message.

yiwu0b11 committed:
cmdstat will set to 3 when error then set to specific value

yi-wu-arm commented:
>but the error message Invalid command line seems not quite accurate.

Correct, but I'm not sure if its a good idea to use `popen` to capture the output (and only store it to `cmdstat` when error occurs)
It would be better if there is a detailed `cmdmsg` to match the actual meaning of the error. but in case of an error, the error message would shows up in the console, maybe adding "Invalid command line: consider checking for exitstat and console printout"

yiwu0b11 committed:
slightly more detailed cmdmsg

DanielCChen commented:
> Correct, but I'm not sure if its a good idea to use `popen` to capture the output (and only store it to `cmdstat` when error occurs)
> It would be better if there is a detailed `cmdmsg` to match the actual meaning of the error. but in case of an error, the error message would shows up in the console, maybe adding "Invalid command line: consider checking for exitstat and console printout"

My intention was to separate the error messages of a failed command execution from the command doesn't exist at all. It is a minor issue though.

yiwu0b11 committed:
Revert "slightly more detailed cmdmsg"

This reverts commit 3b285a1024f3115a6af6da1b0d7d637548a0bf1c.

yiwu0b11 committed:
add cmdmsg to common exit code 1, 126, 127 for Linux

for Windows, all three of them has a exit code of 1
so can't write cmdmsg for them

yi-wu-arm commented:
> separate the error messages

Done, but only on Linux. `cd nowhere` and `fakecommand` would all have a exitstat code of 1 on Windows so I can't seperate them. I thought about using `errno`, but it turns out that `errno` is set to 0 for the above two commands.

yiwu0b11 committed:
rewrite if-else flow

yi-wu-arm commented on flang/runtime/execute.cpp:
```diff
@@ -64,22 +64,40 @@ void CheckAndStoreIntToDescriptor(
+      CheckAndCopyCharsToDescriptor(cmdmsg,
+          "Command not found: command was not found in the system's PATH");
+    }
+  } else
 #endif
+  if (exitStatusVal != 0) {
```
I don't know if there is a prettier way to write this logic:
On Linux (remember the !=0 covers all other error exit code)
```
if exit ==1
else if exit == 126
else if exit == 127
else if exit != 0
```
On Windows
```
if exit != 0
```

So in my code it becomes:
```
#ifdef Linux
  if exit ==1
  else if exit == 126
  else if exit == 127
  else (connect to the following if to beocome a else-if)
#endif 
if exit != 0

```

DanielCChen replied:
Thanks for the update!
It seems to me the "old" logic is

```
if (status == -1) { //  Case 1: common for both Linux and Windows
...
}
#ifdef windows
  if (exitStatusVal == 1) { // Case 2 for all errors on windows.
#else
  if (exitStatusVal == 126 or 127 { // Case 3 (126) and Case 4 (127) for Linux.
#end if
...
}
```

I think what is missing is there is no checking for `exitStatusVal == 1` to catch the case in my issue on Linux. 
I am not sure if we need `exitStatusVal != 0` case though. 
Based on your commits, may I suggest something like:
```
enum CMD_STAT {
  ASYNC_NO_SUPPORT_ERR = -2,
  NO_SUPPORT_ERR = -1,
  CMD_EXECUTED = 0,
  FORK_ERR = 1,
  COMMAND_EXECUTION_ERR = 2, // exit code 1
  COMMAND_CANNOT_EXECUTE_ERR = 3, // Linux exit code 126
  COMMAND_NOT_FOUND_ERR = 4, // Linux exit code 127
  SIGNAL_ERR = 5
};
...

  if (status == -1) {
    if (!cmdstat) {
      terminator.Crash("Command line execution is not supported with system status code: %d.", status);
    } else {
      StoreIntToDescriptor(cmdstat, NO_SUPPORT_ERR, terminator);
      CheckAndCopyCharsToDescriptor(cmdmsg, "Command line execution is not supported.");
    }
  }
#ifdef _WIN32
  // On WIN32 API std::system returns exit status directly
  int exitStatusVal{status};
  if (exitStatusVal == 1) {
    if (!cmdstat) {
      terminator.Crash(
          "Invalid command quit with exit status code: %d.", exitStatusVal);
    } else {
      StoreIntToDescriptor(cmdstat, INVALID_CL_ERR, terminator);
      CheckAndCopyCharsToDescriptor(cmdmsg, "Invalid command line."); // Since Windows doesn't have different return code for different errors, use a generic error message.
    }
  }
#else
  int exitStatusVal{WEXITSTATUS(status)};
  if (exitStatusVal == 1) {
    if (!cmdstat) {
      terminator.Crash("Command line execution failed with exit code: 1.");
    } else {
      StoreIntToDescriptor(cmdstat, COMMAND_EXECUTION_ERR , terminator);
      CheckAndCopyCharsToDescriptor(cmdmsg,
          "Command line execution failed with exit code: 1.");
    }
  } else if (exitStatusVal == 126) {
    if (!cmdstat) {
      terminator.Crash("Command cannot be executed with exit code: 126.");
    } else {
      StoreIntToDescriptor(cmdstat, COMMAND_CANNOT_EXECUTE_ERR, terminator);
      CheckAndCopyCharsToDescriptor(cmdmsg,
          "Command cannot be executed with exit code: 126.");;
    }
  } else if (exitStatusVal == 127) {
    if (!cmdstat) {
      terminator.Crash("Command not found with exit code: 127.");
    } else {
      StoreIntToDescriptor(cmdstat, COMMAND_NOT_FOUND_ERR, terminator);
      CheckAndCopyCharsToDescriptor(cmdmsg,
          "Command not found with exit code: 127.");
    }
  }
  ```
  
  Please note that I am not sure what case `status == -1` covers. I assumed it is for the case that a processor that doesn't support command line execution, so I changed the `CMD_STAT` of it from `EXECL_ERR` to `NO_SUPPORT_ERR`. Please correct me if I am wrong. BTW, 'NO_SUPPORT_ERR` seems is not used anywhere.

yi-wu-arm replied:
Thanks a lot for your inputs, I agree that give specific `CMD_STAT` for common exit code is a good idea.

---

`exitStatusVal != 0` was meant to be a general error message, covers all other error code that wasn't specified by the if-else. Otherwise `CMD_MSG` and `CMD_STAT` would only be set if error code is 1, 126, 127.

---

>BTW, `NO_SUPPORT_ERR` seems is not used anywhere otherwise.  

On Windows
>https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/system-wsystem?view=msvc-170
>A return value of -1 indicates an error, and errno is set to one of the following values  

| Value | Description |
| --- | --- |
| E2BIG | The argument list (which is system-dependent) is too large. |
| ENOENT | The command interpreter can't be found. |
| ENOEXEC | The command-interpreter file can't be executed because the format isn't valid. |
| ENOMEM | Not enough memory is available to execute command; or available memory has been corrupted; or a non-valid block exists, which indicates that the calling process has been allocated incorrectly. |

On Linux: (`std::system` is achieved by `fork()`)
>https://man7.org/linux/man-pages/man3/system.3.html
>If a child process could not be created, or its status could not be retrieved, the return value is -1 and [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to indicate the error.

In other words, a return code of -1 could mean `EXECL_ERR` **or** `NO_SUPPORT_ERR` based on the value of errno, thats why I used `EXECL_ERR`, for safety. On a second thought, I think it would be best to report the value of `errno` as well.

DanielCChen replied:
Ok, I see.
1. For `status == -1` case, I am OK with specifying specific exit code with corresponding error messages depends on the platform.
2. For `exitStatusVal == 1/126/127` case, we have already covered them.
3. For `exitStatusVal != 0` on top of 1 and 2, I am OK with adding another `else` to cover all "other" cases for all platform.

yi-wu-arm replied:
Thanks a lot! I have updated the code based on previous comments.
I have rebased on main, hopefully this will solve some unrelated test failure.... Waiting for CI result.

yiwu0b11 committed:
test fixes Windows

yiwu0b11 committed:
clang format

yi-wu-arm commented:
Win CI build fail: `fatal error C1060: compiler is out of heap space` (build and pass all tests on my local build)
Linux CI build and pass all tests

yi-wu-arm commented:
Did a rebase on main
Hi @DanielCChen do you mind to re-review the code? I've added cmdmsg for common exit code: 127, 126, 1.

Unfortunately, this doesn't work on Windows, so it's an Linux only feature.

yiwu0b11 committed:
add cmdstat and cmdmsg for common linux exit code 1 126 127

yiwu0b11 committed:
test fixes

yi-wu-arm commented:
Sorry for the pin @DanielCChen , do you mind take a l made. In short, added cmdstat for 1, 126, 127, added NO_SUPPORT_ERR based on return value and errno.

DanielCChen commented:
Thanks for the update! It looks good to me. 
The only minor suggestion is to move the code that checks `status == =1` prior to the code that checks `exitStatusVal` because it was the order before, so it would be easier to compare the changes.

yiwu0b11 committed:
move status==-1 back to its original place

DanielCChen reviewed (approved):
LGTM. Thanks for fixing this!

yi-wu-arm commented:
This has cause test failure on the llvm-test-suite, because we provide different cmdstat and more detailed cmdmsg.
Test would fail on:
[execute_command_line_1.f90](https://github.com/llvm/llvm-test-suite/blob/6a38b2969dab9a0c98c6de3c792294f9e118b8de/Fortran/gfortran/regression/execute_command_line_1.f90#L31-L32): (Line31):   `call execute_command_line ("ls *.doesnotexist", .true., i)` should be terminated but the test expected not. (if a non-zero value is assigned to `cmdstat` but `cmdstat` is not present, terminate)
[execute_command_line_3.f90](https://github.com/llvm/llvm-test-suite/blob/6a38b2969dab9a0c98c6de3c792294f9e118b8de/Fortran/gfortran/regression/execute_command_line_3.f90#L14-L15): we provide different cmdstat and better cmdmsg
No affect on:
https://github.com/llvm/llvm-test-suite/blob/6a38b2969dab9a0c98c6de3c792294f9e118b8de/Fortran/gfortran/regression/execute_command_line_2.f90#L4
